### PR TITLE
Twod fitter

### PR DESCRIPTION
**Pull request by matteolucchini1**
@bjricketts It's done :)

### CODE REVIEW

File: docs/fit_twod.ipynb
Comment (bjricketts) [inline docs/fit_twod.ipynb:1]: Reminder to check that this has compiled correctly on the beta readthedocs once this is merged.

File: ndspec/FitPowerSpectrum.py [hunk 1]
```diff
@@ -84,7 +84,6 @@ class FitPowerSpectrum(SimpleFit,FrequencyDependentFit):
     def __init__(self,likelihood="chisq"):
         SimpleFit.__init__(self,likelihood)
         self.freqs = None 
-        self.dependence = "frequency"
```
Comment (bjricketts) [inline ndspec/FitPowerSpectrum.py:87]: Is this inherited from FrequencyDependentFit, or is it just irrelevant now?

File: ndspec/FitTwoD.py [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+from lmfit import Parameters as LM_Parameters
+
+from .Response import ResponseMatrix
+from .SimpleFit import SimpleFit
+from .Likelihoods import chisq, ratio
+
+class FitTwoD(SimpleFit):
+    """
+    This class is designed for fitting generic types of two-dimensional data, 
+    regardless of what units it may be in. Models used in this fitter are
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:22]: Would be good to give an example here of the use case - such as attempting to fit data with respect to some other property of the data (such as phase or polarization degree).

File: ndspec/FitTwoD.py [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+   
+    likelihood: str
+        A string that allows to switch between different fit statistics; which 
+        one is available depends on the type of fitter object. Uses chi-squared 
+        likelihood by default. Users can set different likelihoods either at 
+        initialization or with the appropriate setter method.
+        
+    custom_likelihood: function 
+        A function users can set to bypass the supported likelihoods and instead 
+        provide their own.
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:50]: Currently, someone can't pass a custom likelihood function at initialization. Is this worth doing?

File: ndspec/FitTwoD.py [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+   
+    model_params: lmfit.Parameters 
+        A lmfit Parameters object, which contains the parameters for the model 
+        components.
+   
+    likelihood: str
+        A string that allows to switch between different fit statistics; which 
+        one is available depends on the type of fitter object. Uses chi-squared 
+        likelihood by default. Users can set different likelihoods either at 
+        initialization or with the appropriate setter method.
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:46]: Current behaviour does not line up with this docstring: can't pass a custom likelihood in at initialization.

File: ndspec/FitTwoD.py [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+        -----------
+        bound_lo : float
+            Lower bound of ignored rows interval.
+        bound_hi : float
+            Higher bound of ignored rows interval.    
+        """
+    
+        if ((isinstance(bound_lo, (np.floating, float, int)) != True)|
+            (isinstance(bound_hi, (np.floating, float, int)) != True)):
+            raise TypeError("Energy bounds must be floats or integers")
```
Comment (bjricketts) [inline ndspec/FitTwoD.py]: Should this type error always be "energy bounds"? What if it's frequency or another unit?

File: ndspec/FitTwoD.py [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+        -----------
+        bound_lo : float
+            Lower bound of noticed rows interval.
+        bound_hi : float
+            Higher bound of noticed rows interval.    
+        """
+    
+        if ((isinstance(bound_lo, (np.floating, float, int)) != True)|
+            (isinstance(bound_hi, (np.floating, float, int)) != True)):
+            raise TypeError("Energy bounds must be floats or integers")
```
Comment (bjricketts) [inline ndspec/FitTwoD.py]: Similar comment here. I don't think this should specify energy bounds, as it could be another unit.

File: ndspec/FitTwoD.py [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+            row_grid = self.ear                      
+        
+        if params is None:
+            model = self.model.eval(self.model_params,x_axis=column_grid,y_axis=row_grid)
+        else:
+            model = self.model.eval(params,x_axis=column_grid,y_axis=row_grid)            
+
+        #add folding of the response if necessary here 
+        #the transpositions are necessary because of the weirdness introduce 
+        #by .flatten(). TBD check that this makes sense with xpsec models
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:471]: Make a reminder issue for this please :)

File: ndspec/FitTwoD.py (hunk 4 above)
Comment (bjricketts) [reply ndspec/FitTwoD.py:46]: This should also tell someone to pass "custom" if they are using a custom likelihood, and error if custom is passed without an actual function.

File: ndspec/FitTwoD.py [hunk 8]
```diff
@@ -0,0 +1,677 @@
...
+            ymax = self.ebounds[-1]+0.5*self.ewidths[-1]
+            ax1.set_ylim([ymin,ymax])
+        
+        fig.tight_layout()
+        if return_plot is True:
+            return fig
+        else:
+            return  
+            
+    def plot_model(self,residuals="chisq",params=None,
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:576]: Is it worth allowing the user to decide whether to plot the data or not? Default behaviour to plot the data as well is probably fine.

**Review by bjricketts — CHANGES_REQUESTED** [review]

File: ndspec/FitPowerSpectrum.py (hunk 1 above)
Comment (matteolucchini1) [reply ndspec/FitPowerSpectrum.py:87]: I added this for consistency with the other fitters. As long as you are dealing with one-dimensional data it's not relevant, it's only necessary for 2d stuff (like the generic one or the cross spectrum) to set up masks correctly

File: ndspec/FitTwoD.py (hunk 2 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:22]: https://github.com/nDspec/nDspec/pull/95/changes/c9585ec7a226be294868bab7e7cc4ede81223ff6

File: ndspec/FitTwoD.py (hunk 3 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:50]: hmm passing at initialization is doable but a little annoying. I would prefer if users were very deliberate with custom likelihoods, and set their own, so maybe we just say that if you want a custom likelihood you define it after initialization?

File: ndspec/FitTwoD.py (hunk 8 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:576]: Since users must define their own functions out of the fitter to plot the model anyway, I don't think it's necessary here just like it is not necessary for the cross spectrum fitter.

File: ndspec/FitTwoD.py (hunk 2 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:22]: But also we shouldn't mention polarization, that will require a bit of dedicated functionality and new operators if we're going to make it decently user friendly

File: ndspec/FitTwoD.py (hunk 4 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:46]: https://github.com/nDspec/nDspec/pull/95/changes/e9fcfbe0f48547df0020bc503c20110133d9f10d

I fixed the docstring. I didnt' add raising errors and stuff because much of the beta stuff also doesn't, I will do a final pass on that once we are done with new features (which realistically, is basically after this PR is done).

File: ndspec/FitTwoD.py (hunk 6 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py]: https://github.com/nDspec/nDspec/pull/95/changes/e9fcfbe0f48547df0020bc503c20110133d9f10d fixed

File: ndspec/FitTwoD.py (hunk 7 above)
Comment (matteolucchini1) [reply ndspec/FitTwoD.py:471]: https://github.com/nDspec/nDspec/issues/96

File: docs/fit_twod.ipynb
Comment (matteolucchini1) [reply docs/fit_twod.ipynb:1]: I made a separate build to check this in advance, it's fine

File: ndspec/FitTwoD.py [hunk 9]
```diff
@@ -0,0 +1,679 @@
...
+from .SimpleFit import SimpleFit
+from .Likelihoods import chisq, ratio
+
+class FitTwoD(SimpleFit):
+    """
+    This class is designed for fitting generic types of two-dimensional data, 
+    regardless of what units it may be in. Models used in this fitter are 
+    expected to be provided already in the same unit as the data. Common 
+    examples of using this class might be time-dependent spectroscopy, or 
+    fitting a dynamica power spectrum.
```
Comment (bjricketts) [inline ndspec/FitTwoD.py]: typo: dynamica should be dynamical

File: ndspec/FitTwoD.py [hunk 10]
```diff
@@ -0,0 +1,679 @@
...
+
+class FitTwoD(SimpleFit):
+    """
+    This class is designed for fitting generic types of two-dimensional data, 
+    regardless of what units it may be in. Models used in this fitter are 
+    expected to be provided already in the same unit as the data. Common 
+    examples of using this class might be time-dependent spectroscopy, or 
+    fitting a dynamica power spectrum. 
+    
+    As an exception, users can optionally pass an istrument response matrix
```
Comment (bjricketts) [inline ndspec/FitTwoD.py]: typo: istrument -> instrument

File: ndspec/FitTwoD.py (hunk 3 above)
Comment (bjricketts) [reply ndspec/FitTwoD.py:50]: #97

File: ndspec/FitTwoD.py [hunk 11]
```diff
@@ -0,0 +1,679 @@
...
+            self._noise_unmasked = None
+            self._noise_err_unmasked = None           
+        if self.response is not None:
+            self._emin_unmasked = self.response.emin
+            self._emax_unmasked = self.response.emax
+            self._ebounds_unmasked = self.ebounds
+            self._ewidths_unmasked = self.ewidths
+        return 
+
+    def set_fit_statistic(self,stat):
```
Comment (bjricketts) [inline ndspec/FitTwoD.py:242]: Stats naming gripe: fit statistic and likelihood are two different concepts. Fine to have a fit statistic setter, but then you shouldn't call it likelihood.

**Review by bjricketts — CHANGES_REQUESTED** [review]

**Review by bjricketts — APPROVED** [review]
